webrender fails to build with --default-features=false + features = "pathfinder"

```toml
[dependencies.webrender]
git = "https://github.com/servo/webrender"
rev = "f9ddc5261ad2cd35c60851c4091b682c900dbb0a"
features = ["pathfinder"]
```

Since now both pathfinder and webrender link to freetype, this fails to build with:

```rust
error: multiple packages link to native library `freetype`, but a native library can be linked only once

package `freetype-sys v0.6.0`
    ... which is depended on by `pathfinder_font_renderer v0.1.0 (https://github.com/pcwalton/pathfinder#fcfd157c)`
    ... which is depended on by `webrender v0.57.2 (https://github.com/servo/webrender?rev=f9ddc5261ad2cd35c60851c4091b682c900dbb0a#f9ddc526)`
    ... which is depended on by `azul v0.1.0 (file:///home/felix/Development/azul)`
links to native library `freetype`

package `servo-freetype-sys v4.0.3`
    ... which is depended on by `freetype v0.4.0`
    ... which is depended on by `webrender v0.57.2 (https://github.com/servo/webrender?rev=f9ddc5261ad2cd35c60851c4091b682c900dbb0a#f9ddc526)`
    ... which is depended on by `azul v0.1.0 (file:///home/felix/Development/azul)`
also links to native library `freetype`
```

So, my next course of action was to disable the default freetype feature (also note that with the migration to pathfinder, it shouldn't be a default feature anymore IMO):

```toml
[dependencies.webrender]
git = "https://github.com/servo/webrender"
rev = "f9ddc5261ad2cd35c60851c4091b682c900dbb0a"
default-features = false
features = ["pathfinder"]
```

Which fails to build with a few compile errors:

```rust
   Compiling pathfinder_partitioner v0.1.0 (https://github.com/pcwalton/pathfinder#fcfd157c)
error[E0599]: no method named `add_native_font` found for type `&mut pathfinder_font_renderer::FontContext<webrender_api::FontKey>` in the current scope
   --> /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/glyph_rasterizer.rs:867:27
    |
867 |                 drop(self.add_native_font(&font_key, (*native_font_handle).clone().0));
    |                           ^^^^^^^^^^^^^^^

error[E0609]: no field `0` on type `webrender_api::NativeFontHandle`
   --> /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/glyph_rasterizer.rs:867:54
    |
867 |                 drop(self.add_native_font(&font_key, (*native_font_handle).clone().0));
    |                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `compute_scale` found for type `glyph_rasterizer::FontTransform` in the current scope
   --> /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/platform/unix/font.rs:280:49
    |
280 |         let (x_scale, y_scale) = font.transform.compute_scale().unwrap_or((1.0, 1.0));
    |                                                 ^^^^^^^^^^^^^
    |
   ::: /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/glyph_rasterizer.rs:83:1
    |
83  | pub struct FontTransform {
    | ------------------------ method `compute_scale` not found for this

error[E0599]: no method named `invert_scale` found for type `glyph_rasterizer::FontTransform` in the current scope
   --> /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/platform/unix/font.rs:289:44
    |
289 |             let mut shape = font.transform.invert_scale(x_scale, y_scale);
    |                                            ^^^^^^^^^^^^
    |
   ::: /home/felix/.cargo/git/checkouts/webrender-c3596abe1cf4f320/f9ddc52/webrender/src/glyph_rasterizer.rs:83:1
    |
83  | pub struct FontTransform {
    | ------------------------ method `invert_scale` not found for this

error: aborting due to 4 previous errors

error: Could not compile `webrender`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

rustc 1.25 stable
Debian 9.4 (note: in the pathfinder repo, the `add_native_font` uses macOS Quartz fonts?)

--- a/webrender/src/glyph_rasterizer.rs
+++ b/webrender/src/glyph_rasterizer.rs
@@ -8,6 +8,8 @@ use api::{ColorF, ColorU};
 use api::{FontInstanceFlags, FontInstancePlatformOptions};
 use api::{FontKey, FontRenderMode, FontTemplate, FontVariation};
 use api::{GlyphDimensions, GlyphKey, LayerToWorldTransform, SubpixelDirection};
+#[cfg(feature = "pathfinder")]
+use api::NativeFontHandle;
 #[cfg(any(test, feature = "pathfinder"))]
 use api::DeviceIntSize;
 #[cfg(not(feature = "pathfinder"))]
@@ -128,12 +130,12 @@ impl FontTransform {
         )
     }
 
-    #[cfg(not(feature = "pathfinder"))]
+    #[allow(dead_code)]
     pub fn determinant(&self) -> f64 {
         self.scale_x as f64 * self.scale_y as f64 - self.skew_y as f64 * self.skew_x as f64
     }
 
-    #[cfg(not(feature = "pathfinder"))]
+    #[allow(dead_code)]
     pub fn compute_scale(&self) -> Option<(f64, f64)> {
         let det = self.determinant();
         if det != 0.0 {
@@ -145,7 +147,7 @@ impl FontTransform {
         }
     }
 
-    #[cfg(not(feature = "pathfinder"))]
+    #[allow(dead_code)]
     pub fn pre_scale(&self, scale_x: f32, scale_y: f32) -> Self {
         FontTransform::new(
             self.scale_x * scale_x,
@@ -155,7 +157,7 @@ impl FontTransform {
         )
     }
 
-    #[cfg(not(feature = "pathfinder"))]
+    #[allow(dead_code)]
     pub fn invert_scale(&self, x_scale: f64, y_scale: f64) -> Self {
         self.pre_scale(x_scale.recip() as f32, y_scale.recip() as f32)
     }
@@ -453,6 +455,7 @@ impl GlyphRasterizer {
     #[cfg(feature = "pathfinder")]
     fn add_font_to_pathfinder(&mut self, font_key: &FontKey, template: &FontTemplate) {
         let font_contexts = Arc::clone(&self.font_contexts);
+        eprintln!("add_font_to_pathfinder({:?})", font_key);
         font_contexts.lock_pathfinder_context().add_font(&font_key, &template);
     }
 
@@ -861,10 +864,10 @@ impl AddFont for PathfinderFontContext {
     fn add_font(&mut self, font_key: &FontKey, template: &FontTemplate) {
         match *template {
             FontTemplate::Raw(ref bytes, index) => {
-                drop(self.add_font_from_memory(font_key, bytes.clone(), index));
+                drop(self.add_font_from_memory(&font_key, bytes.clone(), index))
             }
             FontTemplate::Native(ref native_font_handle) => {
-                drop(self.add_native_font(font_key, (*native_font_handle).clone().0));
+                drop(self.add_native_font(&font_key, NativeFontHandleWrapper(native_font_handle)))
             }
         }
     }
@@ -1067,3 +1070,6 @@ fn request_render_task_from_pathfinder(glyph_key: &GlyphKey,
 
     Ok((root_task_id, false))
 }
+
+#[cfg(feature = "pathfinder")]
+pub struct NativeFontHandleWrapper<'a>(pub &'a NativeFontHandle);

--- a/webrender/src/platform/macos/font.rs
+++ b/webrender/src/platform/macos/font.rs
@@ -28,6 +28,8 @@ use core_text::font::{CTFont, CTFontRef};
 use core_text::font_descriptor::{kCTFontDefaultOrientation, kCTFontColorGlyphsTrait};
 use gamma_lut::{ColorLut, GammaLut};
 use glyph_rasterizer::{FontInstance, FontTransform};
+#[cfg(feature = "pathfinder")]
+use glyph_rasterizer::NativeFontHandleWrapper;
 #[cfg(not(feature = "pathfinder"))]
 use glyph_rasterizer::{GlyphFormat, GlyphRasterResult, RasterizedGlyph};
 use internal_types::{FastHashMap, ResourceCacheError};
@@ -727,5 +729,12 @@ impl FontContext {
             format: if bitmap { GlyphFormat::ColorBitmap } else { font.get_glyph_format() },
             bytes: rasterized_pixels,
         })
+    }
+}
+
+#[cfg(feature = "pathfinder")]
+impl<'a> Into<CGFont> for NativeFontHandleWrapper<'a> {
+    fn into(self) -> CGFont {
+        (self.0).0.clone()
     }
 }

--- a/webrender/src/platform/unix/font.rs
+++ b/webrender/src/platform/unix/font.rs
@@ -19,7 +19,11 @@ use freetype::freetype::{FT_LOAD_NO_BITMAP, FT_LOAD_NO_HINTING, FT_LOAD_VERTICAL
 use freetype::freetype::{FT_FACE_FLAG_SCALABLE, FT_FACE_FLAG_FIXED_SIZES};
 use freetype::succeeded;
 use glyph_rasterizer::{FontInstance, GlyphFormat, GlyphRasterResult, RasterizedGlyph};
+#[cfg(feature = "pathfinder")]
+use glyph_rasterizer::NativeFontHandleWrapper;
 use internal_types::{FastHashMap, ResourceCacheError};
+#[cfg(feature = "pathfinder")]
+use pathfinder_font_renderer::freetype as pf_freetype;
 use std::{cmp, mem, ptr, slice};
 use std::cmp::max;
 use std::ffi::CString;
@@ -786,5 +790,13 @@ impl Drop for FontContext {
         unsafe {
             FT_Done_FreeType(self.lib);
         }
+    }
+}
+
+#[cfg(feature = "pathfinder")]
+impl<'a> Into<pf_freetype::FontDescriptor> for NativeFontHandleWrapper<'a> {
+    fn into(self) -> pf_freetype::FontDescriptor {
+        let NativeFontHandleWrapper(font_handle) = self;
+        pf_freetype::FontDescriptor::new(font_handle.pathname.clone().into(), font_handle.index)
     }
 }

